src/Altinn.App.Core/Internal/Process/ProcessEventDispatcher.cs

```diff
@@ -0,0 +1,141 @@
+    private readonly ITaskEvents _taskEvents;
+    private readonly IAppEvents _appEvents;
+    private readonly IEvents _eventsService;
+    private readonly bool _registerWithEventSystem;
+    private readonly ILogger<ProcessEventDispatcher> _logger;
+
+    public ProcessEventDispatcher(
+        IInstance instanceService, 
+        IProcess processClient, 
+        ITaskEvents taskEvents, 
+        IAppEvents appEvents, 
+        IEvents eventsService, 
+        IOptions<AppSettings> appSettings,
+        ILogger<ProcessEventDispatcher> logger)
+    {
+        _instanceService = instanceService;
+        _processClient = processClient;
+        _taskEvents = taskEvents;
+        _appEvents = appEvents;
+        _eventsService = eventsService;
+        _registerWithEventSystem = appSettings.Value.RegisterEventsWithEventsComponent;
+        _logger = logger;
+    }
+
+    /// <inheritdoc/>
+    public async Task<Instance> UpdateProcessAndDispatchEvents(Instance instance, Dictionary<string, string>? prefill, List<InstanceEvent>? events)
+    {
+        await HandleProcessChanges(instance, events, prefill);
+
+        // need to update the instance process and then the instance in case appbase has changed it, e.g. endEvent sets status.archived
+        Instance updatedInstance = await _instanceService.UpdateProcess(instance);
+        await _processClient.DispatchProcessEventsToStorage(updatedInstance, events);
+
+        // remember to get the instance anew since AppBase can have updated a data element or stored something in the database.
+        updatedInstance = await _instanceService.GetInstance(updatedInstance);
+
+        return updatedInstance;
+    }
+
+    /// <inheritdoc/>
+    public async Task RegisterEventWithEventsComponent(Instance instance)
+    {
+        if (_registerWithEventSystem)
+        {
+            try
+            {
+                if (!string.IsNullOrWhiteSpace(instance.Process.CurrentTask?.ElementId))
+                {
+                    await _eventsService.AddEvent($"app.instance.process.movedTo.{instance.Process.CurrentTask.ElementId}", instance);
+                }
+                else if (instance.Process.EndEvent != null)
+                {
+                    await _eventsService.AddEvent("app.instance.process.completed", instance);
+                }
+            }
+            catch (Exception exception)
+            {
+                _logger.LogWarning(exception, "Exception when sending event with the Events component");
+            }
+        }
+    }
+
+    /// <summary>
+    /// Will for each process change trigger relevant Process Elements to perform the relevant change actions.
+    ///
+    /// Each implementation 
+    /// </summary>
+    private async Task HandleProcessChanges(Instance instance, List<InstanceEvent>? events, Dictionary<string, string>? prefill)
+    {
+        if (events != null)
+        {
+            foreach (InstanceEvent processEvent in events)
```

Comment:
InstanceEvent and ProcessEvent is used interchangeably and is a bit confusing. I see the process events is within the InstanceEventType which seems be the root of my confusion.